File: server/ml_model.js
import express from 'express';
import path from 'path';
import fs from 'fs';
import { parse } from 'csv-parse/sync';
import brain from 'brain.js';
import cors from 'cors';  // Enable CORS
import { createCanvas } from 'canvas';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const app = express();
const port = 5000;

app.use(cors());  // Enable CORS for all routes

// Load data
const wasteData = fs.readFileSync(path.join(__dirname, 'waste_data.csv'), 'utf8');
const salesData = fs.readFileSync(path.join(__dirname, 'sales_data.csv'), 'utf8');
const recyclingData = fs.readFileSync(path.join(__dirname, 'recycling_data.csv'), 'utf8');
const inventoryData = fs.readFileSync(path.join(__dirname, 'inventory_data.csv'), 'utf8');

const wasteRecords = parse(wasteData, { columns: true, skip_empty_lines: true });
const salesRecords = parse(salesData, { columns: true, skip_empty_lines: true });
const recyclingRecords = parse(recyclingData, { columns: true, skip_empty_lines: true });
const inventoryRecords = parse(inventoryData, { columns: true, skip_empty_lines: true });

// Merge dataframes
const mergedData = salesRecords.map(sale => {
    const inventory = inventoryRecords.find(item => item['Product Name'] === sale['Product Name']);
    const waste = wasteRecords.find(item => item['Product Name'] === sale['Product Name']);
    const recycling = recyclingRecords.find(item => item['Sale Date'] === sale['Sale Date']);
    return {
        ...sale,
        ...inventory,
        ...waste,
        ...recycling
    };
});

// Convert columns to numeric
mergedData.forEach(record => {
    for (const key in record) {
        if (record[key]) {
            record[key] = parseFloat(record[key].replace(/[^\d.]/g, '')) || 0;
        }
    }
});

// Define features and target
const X = mergedData.map(record => [record['Quantity Sold'], record['Stock Level'], record['Quantity_x'], record['Quantity_y']]);
const y = mergedData.map(record => record['Quantity_x']);

// Prepare data for brain.js
const trainingData = X.map((input, index) => ({
    input,
    output: [y[index]]
}));

// Create and train the neural network
const net = new brain.NeuralNetwork({ hiddenLayers: [10] });

net.train(trainingData, {
    iterations: 20000,
    log: true,
    logPeriod: 1000,
    learningRate: 0.01
});

// Predictions and Mean Squared Error
const predictions = X.map(input => net.run(input));
const mse = predictions.reduce((sum, pred, index) => sum + Math.pow(pred[0] - y[index], 2), 0) / y.length;

// API to fetch predictions and MSE
app.get('/api/waste-predictions', (req, res) => {
    const predictionsData = X.map((input, index) => ({
        actual: y[index],
        predicted: predictions[index][0]
    }));

    res.json({
        predictions: predictionsData,
        mse
    });
});

// Start the server
app.listen(port, () => {
    console.log(`Server is running at http://localhost:${port}`);
});

File: server/ml_model.py
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

# Use Agg backend for matplotlib
import matplotlib
matplotlib.use('Agg')

# Load data
waste_data = pd.read_csv('waste_data.csv')
sales_data = pd.read_csv('sales_data.csv')
recycling_data = pd.read_csv('recycling_data.csv')
inventory_data = pd.read_csv('inventory_data.csv')

# Preprocess data
def convert_to_numeric(column):
    return pd.to_numeric(column.astype(str).str.replace(r'[^\d.]', ''), errors='coerce')

# Merge dataframes on Product Name
merged_data = pd.merge(sales_data, inventory_data, on='Product Name')
merged_data = pd.merge(merged_data, waste_data, on='Product Name', how='left')
merged_data = pd.merge(merged_data, recycling_data, left_on='Sale Date', right_on='Recycling Date', how='left')

# Feature Engineering
# ...existing code...

# Convert columns to numeric
merged_data['Quantity Sold'] = convert_to_numeric(merged_data['Quantity Sold'])
merged_data['Stock Level'] = convert_to_numeric(merged_data['Stock Level'])
merged_data['Quantity_x'] = convert_to_numeric(merged_data['Quantity_x'])
merged_data['Quantity_y'] = convert_to_numeric(merged_data['Quantity_y'])

# Fill missing values
merged_data.fillna(0, inplace=True)

# Split data into features and target
X = merged_data[['Quantity Sold', 'Stock Level', 'Quantity_x', 'Quantity_y']]
y = merged_data['Quantity_x']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Evaluate model
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print(f'Mean Squared Error: {mse}')

# Visualization
plt.figure(figsize=(10, 6))
plt.scatter(y_test, y_pred, alpha=0.5)
plt.xlabel('Actual Waste Quantity')
plt.ylabel('Predicted Waste Quantity')
plt.title('Actual vs Predicted Waste Quantity')
plt.savefig('output_plot.png')
